app-ktor-server/src/main/kotlin/com/ustadmobile/lib/rest/ContentEntryLinkImporter.kt

```diff
@@ -0,0 +1,111 @@
+import com.ustadmobile.core.db.JobStatus
+import com.ustadmobile.core.db.UmAppDatabase
+import com.ustadmobile.core.io.ext.getSize
+import com.ustadmobile.core.util.DiTag
+import com.ustadmobile.core.util.createTemporaryDir
+import com.ustadmobile.door.DoorUri
+import com.ustadmobile.door.ext.DoorTag
+import com.ustadmobile.lib.db.entities.ContentEntryWithLanguage
+import com.ustadmobile.lib.db.entities.ContentJob
+import com.ustadmobile.lib.db.entities.ContentJobItem
+import io.ktor.application.call
+import io.ktor.http.HttpStatusCode
+import io.ktor.request.*
+import io.ktor.response.respond
+import io.ktor.routing.Route
+import io.ktor.routing.post
+import io.ktor.routing.route
+import kotlinx.coroutines.withTimeout
+import org.kodein.di.instance
+import org.kodein.di.ktor.closestDI
+import org.kodein.di.on
+import java.io.File
+import java.net.URL
+
+
+private val IMPORT_LINK_TIMEOUT_DEFAULT = (30 * 1000).toLong()
+
+fun Route.ContentEntryLinkImporter() {
+
+    route("import") {
+
+        post("validateLink") {
+            val url = call.request.queryParameters["url"]?: ""
+            val pluginManager: ContentPluginManager by closestDI().on(call).instance()
+            var metadata: MetadataResult? = null
+            try{
+                val processContext = ProcessContext(createTemporaryDir("content"), mutableMapOf())
+                metadata = withTimeout(IMPORT_LINK_TIMEOUT_DEFAULT) {
+                    pluginManager.extractMetadata(DoorUri.parse(url), processContext)
+                }
+            }catch (e: Exception){
+                Napier.e("Exception extracting metadata to validateLink: $url", e)
+            }
+
+            if (metadata == null) {
+                call.respond(HttpStatusCode.BadRequest, "Unsupported")
+            } else {
+                call.respond(metadata)
+            }
+        }
+
+        post("downloadLink") {
+            try {
+                val parentUid = call.request.queryParameters["parentUid"]?.toLong()
+                val url = call.request.queryParameters["url"]?: ""
+                Napier.i("Downloadlink: $url")
+
+                val contentEntry = call.receive<ContentEntryWithLanguage>()
+                val pluginId = call.request.queryParameters["pluginId"]?.toInt() ?: 0
+                val conversionParams = call.request.queryParameters["conversionParams"]
+                Napier.i("contentEntry from client: ${contentEntry.contentEntryUid}")
+
+                val db: UmAppDatabase by closestDI().on(call).instance(tag = DoorTag.TAG_DB)
+                val repo: UmAppDatabase by closestDI().on(call).instance(tag = DoorTag.TAG_REPO)
+                val containerFolder: File by closestDI().on(call).instance(tag = DiTag.TAG_DEFAULT_CONTAINER_DIR)
+                val entryFromDb = db.contentEntryDao.findByUid(contentEntry.contentEntryUid)
+                val endpoint = Endpoint(call.request.header("Host") ?: "localhost")
+                if (entryFromDb == null) {
```

Comment:
Why do we need to stop here? This looks like a race condition.

Reply:
stop where?